### PR TITLE
feat: LT-1027 - PV Grading

### PR DESCRIPTION
**Pull request by andytaylorkainos**

### CODE REVIEW

File: goods/tests/tests_create.py [hunk 1]
```diff
@@ -1,97 +1,234 @@
...
+    def test_create_good_when_control_code_is_missing_then_bad_request_response_is_returned(self):
+        request_data = _setup_request_data(is_good_controlled=GoodControlled.YES)
+
+        response = self.client.post(url, request_data, **self.exporter_headers)
+
+        self.assertEquals(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEquals(response.json()["errors"], {"control_code": ["This field may not be null."]})
+        self.assertEquals(Good.objects.all().count(), 0)
+
+    def test_create_good_when_control_code_is_invalid_then_bad_request_response_is_returned(self):
```
Comment (BenAndrew-Kainos) [inline goods/tests/tests_create.py]: With all of these tests could we make them parametrised with the request_data & the expected error

File: goods/tests/tests_view.py [hunk 2]
```diff
@@ -11,7 +11,7 @@
 
 class GoodViewTests(DataTestClient):
     def test_view_good_details(self):
-        good = Good(description="thing", is_good_controlled=GoodControlled.NO, organisation=self.organisation,)
+        good = Good(description="thing", is_good_controlled=GoodControlled.NO, organisation=self.organisation)
```
Comment (BenAndrew-Kainos) [inline goods/tests/tests_view.py]: Did black not add the , ?

File: queries/goods_query/helpers.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+from queries.goods_query.models import GoodsQuery
+from static.statuses.enums import CaseStatusEnum
+from static.statuses.libraries.get_case_status import get_case_status_by_status
+
+
+def get_goods_query_by_good(good):
+    try:
+        return GoodsQuery.objects.get(good=good)
+    except GoodsQuery.DoesNotExist:
+        raise NotFoundError({"GoodsQuery": "Goods query not found"})
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/helpers.py]: Move error to content if possible

File: queries/goods_query/helpers.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+
+def is_goods_query_finished(query: GoodsQuery):
+    flags = [
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/helpers.py]: Docstring to why this is needed please

File: queries/goods_query/models.py [hunk 5]
```diff
@@ -6,12 +6,14 @@
...
 
 
-class ControlListClassificationQuery(Query):
+class GoodsQuery(Query):
     """
     Query into getting the correct control list classification for a good
     """
 
-    details = models.TextField(default=None, blank=True, null=True)
+    clc_raised_reasons = models.TextField(default=None, blank=True, null=True)
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/models.py]: Max lengths please

File: queries/goods_query/models.py [hunk 6]
```diff
@@ -6,12 +6,14 @@
...
+    pv_grading_raised_reasons = models.TextField(default=None, blank=True, null=True)
     good = models.ForeignKey(Good, on_delete=models.DO_NOTHING, null=False, related_name="good")
 
-    notifications = GenericRelation(ExporterNotification, related_query_name="clc_query")
+    # TODO: fix related query name
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/models.py]: Please elaborate

File: queries/goods_query/tests.py [hunk 7]
```diff
@@ -4,21 +4,23 @@
...
 from static.statuses.enums import CaseStatusEnum
+from static.statuses.libraries.get_case_status import get_case_status_by_status
 from static.statuses.models import CaseStatus
 from test_helpers.clients import DataTestClient
 from users.models import Role, GovUser
 
 
 class ControlListClassificationsQueryCreateTests(DataTestClient):
-
-    url = reverse("queries:control_list_classifications:control_list_classifications")
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/tests.py]: Why has this replaced the previous tests (do we not need both??)

File: queries/goods_query/tests.py [hunk 8]
```diff
@@ -182,12 +182,144 @@ def test_user_cannot_respond_to_clc_with_case_in_terminal_state(self):
...
         response = self.client.put(url, data, **self.gov_headers)
+        query.refresh_from_db()
+
+        self.assertEqual(query.status.status, CaseStatusEnum.WITHDRAWN)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+
+class PvGradingQueryCreateTests(DataTestClient):
+    def test_given_an_unsure_pv_graded_good_exists_when_creating_pv_grading_query_then_201_created_is_returned(self):
+        pv_graded_good = self.create_good(
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/tests.py]: Move to setup and modify is_pv_graded where applicable

Comment (BenAndrew-Kainos) [reply queries/goods_query/tests.py]: Even if it's just the args in a dict

File: queries/goods_query/tests.py [hunk 9]
```diff
@@ -182,12 +182,144 @@ def test_user_cannot_respond_to_clc_with_case_in_terminal_state(self):
...
+            description="This is a good",
+            org=self.organisation,
+            is_good_controlled=GoodControlled.NO,
+            control_code="",
+            is_pv_graded=GoodPvGraded.GRADING_REQUIRED,
+        )
+        data = {
+            "good_id": pv_graded_good.id,
+            "pv_grading_raised_reasons": "This is the reason why I'm unsure...",
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/tests.py]: Please move reason to a separate variable to use here & in assert

File: queries/goods_query/tests.py [hunk 10]
```diff
@@ -182,12 +182,144 @@ def test_user_cannot_respond_to_clc_with_case_in_terminal_state(self):
...
+        }
+        url = reverse("queries:goods_queries:goods_queries")
 
+        response = self.client.post(url, data, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(
+            response.json()["errors"],
+            [
+                f'A good must have either "is_good_controlled" set to "{GoodControlled.UNSURE}" '
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/tests.py]: Please use string from lite_content

File: queries/goods_query/tests.py [hunk 11]
```diff
@@ -182,12 +182,144 @@ def test_user_cannot_respond_to_clc_with_case_in_terminal_state(self):
...
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
         self.assertEqual(query.status.status, CaseStatusEnum.WITHDRAWN)
+
+
+class CombinedPvGradingAndClcQuery(DataTestClient):
+    def setUp(self):
+        super().setUp()
+
+        role = Role(name="review_goods")
+        role.permissions.set([constants.GovPermissions.REVIEW_GOODS.name])
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/tests.py]: Is there a test where they don't have the permission?

File: queries/goods_query/tests.py [hunk 12]
```diff
@@ -182,12 +182,144 @@ def test_user_cannot_respond_to_clc_with_case_in_terminal_state(self):
...
+            clc_raised_reasons="some clc reasons",
+            pv_grading_raised_reasons="some pv reasons",
+            good=self.pv_graded_and_controlled_good,
+            organisation=self.organisation,
+            type=CaseTypeEnum.GOODS_QUERY,
+            status=get_case_status_by_status(CaseStatusEnum.SUBMITTED),
+        )
+        self.clc_and_pv_query.flags.set(
+            [
+                Flag.objects.get(id=SystemFlags.GOOD_CLC_QUERY_ID),
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/tests.py]: Can you not set the flags in the create?

File: queries/goods_query/views.py [hunk 13]
```diff
@@ -14,51 +12,80 @@
...
 
-class ControlListClassificationsList(APIView):
+class GoodsQueriesCreate(APIView):
     authentication_classes = (ExporterAuthentication,)
 
+    @staticmethod
+    def _check_request_for_errors(good: Good, is_clc_required: bool, is_pv_grading_required: bool):
+        errors = []
+        if good.status != GoodStatus.DRAFT:
+            errors += [{"status": f'A good must have its status set to "{GoodStatus.DRAFT}" to raise a Goods Query"'}]
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/views.py]: Please move to lite_content

File: queries/goods_query/views.py [hunk 14]
```diff
@@ -14,51 +12,80 @@
...
+
+        if not (is_clc_required or is_pv_grading_required):
+            errors += [
+                f'A good must have either "is_good_controlled" set to "{GoodControlled.UNSURE}" '
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/views.py]: lite_content

File: queries/goods_query/views.py [hunk 15]
```diff
@@ -14,51 +12,80 @@
...
+from queries.goods_query.helpers import is_goods_query_finished
+from queries.goods_query.models import GoodsQuery
 from queries.helpers import get_exporter_query
 from static.statuses.enums import CaseStatusEnum
 from static.statuses.libraries.get_case_status import get_case_status_by_status
 from users.models import UserOrganisationRelationship
 
 
-class ControlListClassificationsList(APIView):
+class GoodsQueriesCreate(APIView):
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/views.py]: Docstrings

File: queries/goods_query/views.py [hunk 16]
```diff
@@ -14,51 +12,80 @@
...
         good.save()
-        clc_query.save()
+        goods_query.save()
 
-        return JsonResponse(data={"id": clc_query.id}, status=status.HTTP_201_CREATED)
+        return JsonResponse(data={"id": goods_query.id}, status=status.HTTP_201_CREATED)
 
 
-class ControlListClassificationDetail(APIView):
+class GoodQueryCLCResponse(APIView):
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/views.py]: Docstrings

File: queries/goods_query/views.py [hunk 17]
```diff
@@ -123,7 +151,7 @@ def put(self, request, pk):
         return JsonResponse(data={"errors": clc_good_serializer.errors}, status=status.HTTP_400_BAD_REQUEST)
 
 
-class CLCManageStatus(APIView):
+class GoodQueryManageStatus(APIView):
```
Comment (BenAndrew-Kainos) [inline queries/goods_query/views.py]: Docstrings

File: static/private_venture_gradings/urls.py [hunk 18]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path
+
+from static.private_venture_gradings.views import PVGradingsList
+
+app_name = "private_venture_gradings"
+
+urlpatterns = [
+    path("", PVGradingsList.as_view(), name="private_venture_gradings"),
+    # path("<str:rating>/", views.ControlListEntryDetail.as_view(), name="control_list_entry",),
```
Comment (BenAndrew-Kainos) [inline static/private_venture_gradings/urls.py]: Please remove this and the class if unused

File: static/private_venture_gradings/views.py [hunk 19]
```diff
@@ -0,0 +1,9 @@
+from django.http import JsonResponse
+from rest_framework.views import APIView
+
+from goods.enums import PvGrading
+
+
+class PVGradingsList(APIView):
+    def get(self, request):
+        return JsonResponse(data={"pv-gradings": PvGrading.choices})
```
Comment (BenAndrew-Kainos) [inline static/private_venture_gradings/views.py]: Could you convert choices to a dict before passing (will probably be more useful)

File: Dockerfile [hunk 20]
```diff
@@ -1,7 +1,6 @@
 FROM python:3.7-slim
 WORKDIR /app
-RUN apt-get update --fix-missing
-RUN apt-get install -y libpq-dev gcc \
+RUN apt-get update --fix-missing && apt-get install -y libpq-dev gcc \
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply Dockerfile]: Wasnt me, think mark marich was doing some cleanup.
Ill revert it

File: goods/models.py [hunk 21]
```diff
@@ -5,17 +5,37 @@
...
+from goods.enums import GoodStatus, GoodControlled, PvGrading, GoodPvGraded
 from organisations.models import Organisation
 from static.missing_document_reasons.enums import GoodMissingDocumentReasons
 from users.models import ExporterUser
 
 
+class PvGradingDetails(models.Model):
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    grading = models.CharField(choices=PvGrading.choices, default=None, blank=True, null=True, max_length=30)
+    custom_grading = models.TextField(default="", blank=True, null=True)
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/models.py]: Good question, ill check with Anne

File: goods/serializers.py [hunk 22]
```diff
@@ -4,22 +4,57 @@
...
+            "suffix",
+            "issuing_authority",
+            "reference",
+            "date_of_issue",
+        )
+
+    def validate(self, data):
+        validated_data = super(GoodPvGradingDetailsSerializer, self).validate(data)
+
+        if not validated_data.get("grading") and not validated_data.get("custom_grading"):
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/serializers.py]: I dont think we're allowed to make that assumption - because they could select a grading and also provide a custom one. In which case, we would need to make them be explicit

File: goods/serializers.py [hunk 23]
```diff
@@ -138,15 +185,54 @@ def validate(self, value):
...
+        pv_grading_details = validated_data.pop("pv_grading_details", None)
+
+        if instance:
+            # Only update pv_grading_details if data was provided
+            if pv_grading_details:
+                pv_grading_details = GoodPvGradingDetailsSerializer.update(
+                    GoodPvGradingDetailsSerializer(), validated_data=pv_grading_details, instance=instance,
+                )
+            else:
+                pv_grading_details = instance
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/serializers.py]: if you wouldnt mind checking the refactored version of this code, it'd be appreciated :)

File: goods/tests/tests_create.py [hunk 24]
```diff
@@ -1,97 +1,234 @@
...
+        response = self.client.post(url, request_data, **self.exporter_headers)
 
-    url = reverse("goods:goods")
+        self.assertEquals(response.status_code, status.HTTP_201_CREATED)
+        _assert_response_data(self, response.json()["good"], request_data)
+        self.assertEquals(Good.objects.all().count(), 1)
 
-    @parameterized.expand(
-        [
-            ("Widget", GoodControlled.YES, "ML1a", "1337",),  # Create a new good successfully
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/tests/tests_create.py]: Im fairly certain i translated all the parameterised tests into individual tests or removed them because the same logic was covered in other tests. I will double check tho

File: goods/tests/tests_create.py [hunk 25]
```diff
@@ -1,97 +1,234 @@
...
+
+    def test_create_good_when_is_good_controlled_field_is_missing_then_bad_request_response_is_returned(self):
+        request_data = _setup_request_data()
+        request_data.pop("is_good_controlled")
+
+        response = self.client.post(url, request_data, **self.exporter_headers)
+
+        self.assertEquals(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEquals(
+            response.json()["errors"], {"is_good_controlled": ["Select a value"]},
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/tests/tests_create.py]: Good point!

File: goods/tests/tests_view.py (hunk 2 above)
Comment (andytaylorkainos) [reply goods/tests/tests_view.py]: it did, but only because it used to be on multiple lines

File: queries/goods_query/models.py (hunk 6 above)
Comment (andytaylorkainos) [reply queries/goods_query/models.py]: @Mark-Spiers-kainos spiers

File: cases/tests/test_search_cases.py [hunk 26]
```diff
@@ -89,7 +89,7 @@ def test_get_clc_type_cases(self):
         """
 
         # Arrange
-        url = self.url + "?case_type=clc_query"
+        url = self.url + "?case_type=goods_query"
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply cases/tests/test_search_cases.py]: Done!

File: goods/libraries/save_good.py [hunk 27]
```diff
@@ -0,0 +1,18 @@
+from django.http import JsonResponse
+from rest_framework import status
+
+from conf.helpers import str_to_bool
+
+
+def create_or_update_good(serializer, data, success_response_code):
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/libraries/save_good.py]: Done!

File: queries/goods_query/helpers.py (hunk 4 above)
Comment (andytaylorkainos) [reply queries/goods_query/helpers.py]: When responding to the CLC query or the PV grading on a case that has both queries, we need to be able to ascertain if the case status can now be updated as both queries must be resolved before it can be updated

File: cases/tests/tests_managing_flags_on_case.py [hunk 28]
```diff
@@ -9,7 +9,7 @@ def setUp(self):
         super().setUp()
 
         # Cases
-        self.case = self.create_clc_query("Query", self.organisation)
+        self.case = self.create_end_user_advisory_case("note", "test", self.organisation)
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply cases/tests/tests_managing_flags_on_case.py]: Done!

File: goods/tests/tests_create.py [hunk 29]
```diff
@@ -1,97 +1,234 @@
...
 from rest_framework import status
 from rest_framework.reverse import reverse
 
-from goods.enums import GoodControlled
+from goods.enums import GoodControlled, GoodPvGraded, PvGrading, GoodStatus
+from goods.models import Good
 from test_helpers.clients import DataTestClient
-from test_helpers.decorators import none_param_tester
 
+url = reverse("goods:goods")
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/tests/tests_create.py]: it's used by all of them tho, it'd be duplicated in all classes

File: goods/tests/tests_create.py [hunk 30]
```diff
@@ -1,97 +1,234 @@
...
+def _setup_request_data(
+    is_good_controlled=GoodControlled.NO,
+    control_code=None,
+    is_pv_graded=GoodPvGraded.NO,
+    pv_grading_details=None,
+    part_number="1337",
+    validate_only=False,
+):
+    return {
+        "description": "Plastic bag " + str(uuid.uuid4()),
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/tests/tests_create.py]: Done!

File: goods/tests/tests_create.py [hunk 31]
```diff
@@ -1,97 +1,234 @@
...
+        self.assertEquals(response_data["is_pv_graded"]["key"], request_data["is_pv_graded"])
+
+        pv_grading_details = response_data["pv_grading_details"]
+
+        if request_data["pv_grading_details"]["grading"]:
+            self.assertEquals(pv_grading_details["grading"]["key"], request_data["pv_grading_details"]["grading"])
+
+        self.assertEquals(pv_grading_details["custom_grading"], request_data["pv_grading_details"]["custom_grading"])
+        self.assertEquals(pv_grading_details["prefix"], request_data["pv_grading_details"]["prefix"])
+        self.assertEquals(pv_grading_details["suffix"], request_data["pv_grading_details"]["suffix"])
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/tests/tests_create.py]: Tried doing a for loop but because some of the fields are dynamic depending on what is being tested, and some of those fields also contain nested fields it became fairly convoluted so i decided to keep the fields as individual assertions, thankfully it makes it easier to debug

File: queries/goods_query/models.py (hunk 5 above)
Comment (andytaylorkainos) [reply queries/goods_query/models.py]: It's the same as before just renamed the field

File: goods/tests/tests_create.py (hunk 1 above)
Comment (andytaylorkainos) [reply goods/tests/tests_create.py]: I prefer to keep those types of tests separated to make them easier to debug and understand.
Parameterized testing, for me, makes more sense when youre only making minute changes such as providing an incorrectly formatted field to test validation

File: goods/tests/tests_create.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/tests/tests_create.py]: Actually, these are django errors 😓

File: queries/goods_query/helpers.py (hunk 3 above)
Comment (andytaylorkainos) [reply queries/goods_query/helpers.py]: Done

File: static/private_venture_gradings/urls.py (hunk 18 above)
Comment (andytaylorkainos) [reply static/private_venture_gradings/urls.py]: Done!

File: queries/goods_query/views.py (hunk 17 above)
Comment (andytaylorkainos) [reply queries/goods_query/views.py]: Done!

File: queries/goods_query/views.py (hunk 16 above)
Comment (andytaylorkainos) [reply queries/goods_query/views.py]: Done!

File: queries/goods_query/views.py (hunk 15 above)
Comment (andytaylorkainos) [reply queries/goods_query/views.py]: Done!

File: queries/goods_query/tests.py (hunk 11 above)
Comment (andytaylorkainos) [reply queries/goods_query/tests.py]: yep! "test_user_cannot_respond_to_clc_without_permissions"

File: queries/goods_query/tests.py (hunk 12 above)
Comment (andytaylorkainos) [reply queries/goods_query/tests.py]: Sadly not "TypeError: Direct assignment to the forward side of a many-to-many set is prohibited. Use flags.set() instead."

File: queries/goods_query/views.py (hunk 14 above)
Comment (andytaylorkainos) [reply queries/goods_query/views.py]: Done

File: queries/goods_query/views.py (hunk 13 above)
Comment (andytaylorkainos) [reply queries/goods_query/views.py]: Done

File: queries/goods_query/tests.py (hunk 10 above)
Comment (andytaylorkainos) [reply queries/goods_query/tests.py]: Done

File: queries/goods_query/tests.py (hunk 7 above)
Comment (andytaylorkainos) [reply queries/goods_query/tests.py]: I've updated the CLC tests (moved them to goods_queries) and included new tests for PV gradings. I didnt remove any tests unless the edge-cases were covered elsewhere

File: queries/goods_query/tests.py (hunk 9 above)
Comment (andytaylorkainos) [reply queries/goods_query/tests.py]: Done!

File: goods/serializers.py [hunk 32]
```diff
@@ -99,34 +140,40 @@ def __init__(self, *args, **kwargs):
...
-                "value": get_status_value_from_case_status_enum(clc_query.status.status),
+                "key": goods_query.status.status,
+                "value": get_status_value_from_case_status_enum(goods_query.status.status),
             }
-        except ControlListClassificationQuery.DoesNotExist:
+        except GoodsQuery.DoesNotExist:
             return None
 
     def get_documents(self, instance):
+        if not isinstance(instance, Good):
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/serializers.py]: added them to all

File: goods/serializers.py [hunk 33]
```diff
@@ -99,34 +140,40 @@ def __init__(self, *args, **kwargs):
             if hasattr(self, "initial_data"):
                 self.initial_data["control_code"] = None
 
-    # pylint: disable=W0703
     def get_case_id(self, instance):
-        clc_query = ControlListClassificationQuery.objects.filter(good=instance)
-        if clc_query:
-            return clc_query.first().id
+        if isinstance(instance, Good):
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/serializers.py]: This can happen due to 'validate_only' which doesnt create an instance of the good

File: static/private_venture_gradings/views.py (hunk 19 above)
Comment (andytaylorkainos) [reply static/private_venture_gradings/views.py]: @Mark-Spiers-kainos can we change this on the front-end also please? it'll be:

`response = {"pv_grading": {"key_1": "value_1, "key_2": "value_2"}`

Comment (andytaylorkainos) [reply static/private_venture_gradings/views.py]: Done!

File: queries/goods_query/models.py [hunk 34]
```diff
@@ -6,12 +6,13 @@
...
-class ControlListClassificationQuery(Query):
+class GoodsQuery(Query):
     """
     Query into getting the correct control list classification for a good
     """
 
-    details = models.TextField(default=None, blank=True, null=True)
+    clc_raised_reasons = models.TextField(default=None, blank=True, null=True)
+    pv_grading_raised_reasons = models.TextField(default=None, blank=True, null=True)
     good = models.ForeignKey(Good, on_delete=models.DO_NOTHING, null=False, related_name="good")
```
Comment (efahy) [inline queries/goods_query/models.py]: I can kinda understand why the foreign key is pointing from the query to the good, but it means that the DB would allow for multiple queries to point to the same good.  Could we/should we validate at any point that this is not the case?

File: goods/models.py [hunk 35]
```diff
@@ -5,17 +5,37 @@
...
 from documents.models import Document
 from flags.models import Flag
-from goods.enums import GoodStatus, GoodControlled
+from goods.enums import GoodStatus, GoodControlled, PvGrading, GoodPvGraded
 from organisations.models import Organisation
 from static.missing_document_reasons.enums import GoodMissingDocumentReasons
 from users.models import ExporterUser
 
 
+class PvGradingDetails(models.Model):
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/models.py]: I want to distinguish between the actual PV Grading being assigned (which in this model is the grading field) and the other fields accompanying it.
Therefore this model shows not only the PV Grading, but also the details associated with it

File: goods/models.py [hunk 36]
```diff
@@ -5,17 +5,37 @@
...
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    grading = models.CharField(choices=PvGrading.choices, default=None, blank=True, null=True, max_length=30)
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/models.py]: Done!

File: goods/models.py [hunk 37]
```diff
@@ -5,17 +5,37 @@
...
+    custom_grading = models.TextField(default="", blank=True, null=True, max_length=100)
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/models.py]: Removed them!

File: goods/serializers.py [hunk 38]
```diff
@@ -99,37 +139,47 @@ def __init__(self, *args, **kwargs):
             if hasattr(self, "initial_data"):
                 self.initial_data["control_code"] = None
 
-    # pylint: disable=W0703
+        # This removes data being passed forward from product grading forms on editing goods when not needed
+        if not self.get_initial().get("is_pv_graded") == GoodControlled.YES:
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/serializers.py]: 👀 nicely spotted!

File: goods/serializers.py [hunk 39]
```diff
@@ -138,15 +188,66 @@ def validate(self, value):
...
+            is_pv_graded=instance.is_pv_graded == GoodPvGraded.YES,
+            pv_grading_details=validated_data.get("pv_grading_details"),
+            instance=instance.pv_grading_details,
+        )
+
         instance.save()
         return instance
 
+    @staticmethod
+    def _create_update_or_delete_pv_grading_details(is_pv_graded=False, pv_grading_details=None, instance=None):
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/serializers.py]: I'd like to leave it where it is, so that the update method isn't cluttered if thats okay

File: queries/goods_query/helpers.py [hunk 40]
```diff
@@ -0,0 +1,29 @@
...
+        raise NotFoundError({"GoodsQuery": strings.GoodsQuery.QUERY_NOT_FOUND_ERROR})
+
+
+def update_goods_query_status(query: GoodsQuery):
+    """
+    Ascertain if the Goods Query status should be updated to Finalised or Submitted
+    """
+    flags = [
+        Flag.objects.get(id=SystemFlags.GOOD_CLC_QUERY_ID).id,
+        Flag.objects.get(id=SystemFlags.GOOD_PV_GRADING_QUERY_ID).id,
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply queries/goods_query/helpers.py]: Fixed!

File: goods/serializers.py [hunk 41]
```diff
@@ -4,22 +4,56 @@
...
-from queries.control_list_classifications.models import ControlListClassificationQuery
+from queries.goods_query.models import GoodsQuery
 from static.missing_document_reasons.enums import GoodMissingDocumentReasons
 from static.statuses.libraries.get_case_status import get_status_value_from_case_status_enum
-from users.libraries.get_user import get_user_by_pk
 from users.models import ExporterUser
 from users.serializers import ExporterUserSimpleSerializer
 
 
+class GoodPvGradingDetailsSerializer(serializers.ModelSerializer):
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/serializers.py]: it does, gov!

File: goods/tests/tests_create.py [hunk 42]
```diff
@@ -1,97 +1,228 @@
...
+
+        response = self.client.post(url, request_data, **self.exporter_headers)
+
+        self.assertEquals(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEquals(
+            response.json()["errors"], {"reference": ["This field may not be blank."]},
+        )
+        self.assertEquals(Good.objects.all().count(), 0)
+
+    def test_create_good_when_date_is_missing_then_bad_response_is_returned(self):
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/tests/tests_create.py]: Fixed!

File: goods/tests/tests_create.py [hunk 43]
```diff
@@ -1,97 +1,228 @@
...
+            is_good_controlled=GoodControlled.NO, is_pv_graded=GoodPvGraded.NO, validate_only=True
+        )
+
+        response = self.client.post(url, request_data, **self.exporter_headers)
+
+        self.assertEquals(response.status_code, status.HTTP_200_OK)
+        self.assertEquals(Good.objects.all().count(), 0)
+
+
+class GoodsCreateControlledGoodTests(DataTestClient):
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/tests/tests_create.py]: Done! (in query tests)

File: queries/goods_query/tests.py [hunk 44]
```diff
@@ -182,12 +183,142 @@ def test_user_cannot_respond_to_clc_with_case_in_terminal_state(self):
...
+
+        response = self.client.post(url, data, **self.exporter_headers)
+        response_data = response.json()
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(response_data["id"], str(GoodsQuery.objects.get().id))
+        self.assertEqual(GoodsQuery.objects.count(), 1)
+        self.assertEqual(GoodsQuery.objects.get().pv_grading_raised_reasons, pv_grading_raised_reasons)
+
+    def test_given_a_pv_graded_good_exists_when_creating_pv_grading_query_then_400_bad_request_is_returned(self):
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply queries/goods_query/tests.py]: Done!

File: goods/tests/tests_create.py [hunk 45]
```diff
@@ -1,97 +1,234 @@
...
+        "is_good_controlled": is_good_controlled,
+        "control_code": control_code,
+        "is_pv_graded": is_pv_graded,
+        "pv_grading_details": pv_grading_details,
+        "part_number": part_number,
+        "validate_only": validate_only,
+    }
+
+
+def _setup_pv_grading_details(
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/tests/tests_create.py]: These functions are used to build the `post` which may or may not contain some of the fields.
The other functions in DataTestClient create the object in the DB which is fine when testing get requests (but I need the raw data to test the post endpoints)

File: goods/tests/tests_create.py [hunk 46]
```diff
@@ -1,97 +1,234 @@
...
-from goods.enums import GoodControlled
+from goods.enums import GoodControlled, GoodPvGraded, PvGrading, GoodStatus
+from goods.models import Good
 from test_helpers.clients import DataTestClient
-from test_helpers.decorators import none_param_tester
 
+url = reverse("goods:goods")
+
+
+def _setup_request_data(
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply goods/tests/tests_create.py]: Considering its only used to test the `POST` endpoint for creation of a good, i think it wouldnt make sense to move it to a shared function.

This file contains a lot of classes (so I cant make it class specific)

File: queries/goods_query/models.py (hunk 34 above)
Comment (andytaylorkainos) [reply queries/goods_query/models.py]: Tbh i cant believe that this wasnt already tested - added a test for it!

File: goods/serializers.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply goods/serializers.py]: So long as you enforce that they've chosen 'Other' when entering a custom :)

File: goods/serializers.py (hunk 33 above)
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply goods/serializers.py]: leave a comment somewhere to explain to dumb people like me

File: goods/tests/tests_create.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply goods/tests/tests_create.py]: I mean in a setUp()

File: goods/tests/tests_create.py (hunk 46 above)
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply goods/tests/tests_create.py]: Please change to default dictionary and do a copy & update properties where applicable

File: goods/tests/tests_create.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply goods/tests/tests_create.py]: just move lines 69-76 into the for loop for simplicity

File: goods/serializers.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply goods/serializers.py]: So long as you enforce that they've chosen 'Other' when entering a custom :)

File: goods/serializers.py [hunk 47]
```diff
@@ -99,37 +139,47 @@ def __init__(self, *args, **kwargs):
...
-            return GovUserSimpleSerializer(user).data
+        if self.goods_query_case:
+            return GovUserSimpleSerializer(self.goods_query_case.case_officer).data
+        return None
 
     def get_query(self, instance):
-        return get_good_query_with_notifications(
-            good=instance, exporter_user=self.context.get("exporter_user"), total_count=False
-        )
+        if isinstance(instance, Good):
```
Comment (BenAndrew-Kainos) [inline goods/serializers.py]: leave a comment somewhere to explain to dumb people like me why this is needed

File: goods/tests/tests_create.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply goods/tests/tests_create.py]: Please move to test setup()

File: queries/goods_query/models.py (hunk 5 above)
Comment (BenAndrew-Kainos) [reply queries/goods_query/models.py]: Please follow up about max length here

File: cases/tests/tests_search_cases.py [hunk 48]
```diff
@@ -89,7 +90,7 @@ def test_get_clc_type_cases(self):
         """
 
         # Arrange
-        url = self.url + "?case_type=clc_query"
+        url = self.url + "?case_type=" + CaseTypeEnum.GOODS_QUERY
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply cases/tests/tests_search_cases.py]: Done!

File: goods/serializers.py (hunk 47 above)
Comment (andytaylorkainos) [reply goods/serializers.py]: Done!
